src/umpire/util/AllocationRecordMap.cpp

```diff
@@ -0,0 +1,387 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory
+//
+// Created by David Beckingsale, david@llnl.gov
+// LLNL-CODE-747640
+//
+// All rights reserved.
+//
+// This file is part of Umpire.
+//
+// For details, see https://github.com/LLNL/Umpire
+// Please also see the LICENSE file for MIT license.
+//////////////////////////////////////////////////////////////////////////////
+#include "umpire/util/AllocationRecordMap.hpp"
+
+#include "umpire/util/FixedMallocPool.hpp"
+
+#include "umpire/util/Macros.hpp"
+
+#include <iterator>
+#include <sstream>
+
+namespace umpire {
+namespace util {
+
+namespace {
+
+template <typename T>
+struct ListBlock
+{
+  T rec;
+  ListBlock* prev;
+};
+
+static umpire::util::FixedMallocPool block_pool(sizeof(ListBlock<AllocationRecord>));
+
+// Forward declare const iterator
+class RecordListConstIterator;
+
+// TODO Profile this to determine if we should add a small string
```

Comment:
Should this TODO be a jira ticket?

Reply:
Yes, it could be. I dislike having to be VPNed to access the JIRA, though. Let's see how the performance is after merging. My thinking is that having a small array of RecordLists will be no faster (or slower) than what is now implemented here with the PR, since we have to add an additional integer and keep track of how many elements of the RecordListArray are set/used.

Reply:
Are we merging before running performance tests?  My vote is that we run performance tests (and possibly make updates) before we merge.

Reply:
In this case we can probably quickly assess that before merging, yes. In general my vote is to keep PRs small and keep merging. Merge this, have another PR that expands the benchmark suite, check if there's room for improvement, and iterate.